clients/src/main/java/org/apache/kafka/clients/consumer/internals/RequestState.java

```diff
@@ -106,6 +106,13 @@ public void onSendAttempt(final long currentTimeMs) {
         this.lastSentMs = currentTimeMs;
     }
 
+    /**
+     * Update the lastReceivedTime in milliseconds, indicating that a response has been received.
+     */
+    public void updateLastReceivedTime(final long lastReceivedMs) {
+        this.lastReceivedMs = lastReceivedMs;
```

Comment:
I find doing this a little strange because we ends up in a state where the last received time is updated but the backoff and the number of attempts are not. Therefore, we could still have an non-desired backoff applied to the next request. For instance, imagine having a request timeout followed by a coordinator not available error. Is this case, the backoff is not reset. Is my understanding correct?

Reply:
You're probably right @dajac. Should we be applying exponential backoff for heartbeats? Does it only apply to a subset of errors? As a starting position, I would have said that any response or error updates the last receive time, and we would apply exponential backoff in all cases.

Reply:
Thanks for the feedback. Agree with @dajac's point that we might be carrying on with a non-desired backoff by leaving it unchanged in this situation, so we definitely need to be setting one, and as @AndrewJSchofield said, all errors would update the last receive time and apply "some" backoff. 

That leads to deciding which one, and my intention was 0 backoff in a subset of errors. As I see it, there are some specific errors in HB where we don't  just retry the same request, but rather do a different one, so it makes sense to skip backoff as an optimization, ex. 
1. HB to rejoin as new member after fence error. We would send it right away after the fence, skipping backoff
2. HB to a new coordinator after not_coordinator error. We would send it as soon as the new coordinator is discovered, without applying any backoff.

Makes sense? I will update to make sure we always set a backoff, but with custom 0 backoff for these cases as the initial intention was, if it makes sense.

Reply:
The implementation of exponential backoff for KIP-580 introduced the idea of "equivalent responses". The idea is that exponential backoff applies when metadata responses are "equivalent", but when things are progressing, the backoff is avoided. I think a similar principle could be helpful here. Essentially, if "progress" is being made such as a coordinator change, do not apply the backoff.

Reply:
Exactly, I did see the skip backoff logic in produce when a new leader is discovered, related to what you described. I think that the concept of "progress" here would be more abstract but still applicable, depending on the exact error we know that the action it triggers is based on some progress (send HB as new member, send HB when coord available)